### PR TITLE
clean up backend types and return values

### PR DESCRIPTION
**Pull request by mplsgrant**
### Issue
Getting type errors when running PyCharm's automated checker. Some related to #371.

### Cause
The inconsistent types and return values exist but do not trigger any tests.

### Solution
Update the types and return values as appropriate.

### CODE REVIEW

File: src/backends/compose/compose_backend.py [hunk 1]
```diff
@@ -177,18 +177,18 @@ def get_file(self, tank_index: int, service: ServiceType, file_path: str):
...
 
-    def get_tank_ip_addr(self, index: int) -> str:
-        self.get_tank_ipv4(index)
+    def get_tank_ip_addr(self, index: int) -> str | None:
+        return self.get_tank_ipv4(index)
 
-    def get_tank_dns_addr(self, index: int) -> str:
-        self.get_tank_ip_addr(index)
+    def get_tank_dns_addr(self, index: int) -> str | None:
+        return self.get_tank_ip_addr(index)
```
Comment (pinheadmz) [inline src/backends/compose/compose_backend.py:191]: crap how'd we miss this? how did this pass the tests?

**Review by pinheadmz — COMMENTED** [review]
LGTM, should we add PyCharm to our lint tests?

File: src/backends/compose/compose_backend.py (hunk 1 above)
Comment (mplsgrant) [reply src/backends/compose/compose_backend.py:191]: The tests never call that function!

The other thing I noticed was that the dag_connect test does not call `debug_assert_log`. I think that means that as long as the test runs, it passes. Need to stitch the tests up.

**Comment by mplsgrant** [comment]
> LGTM, should we add PyCharm to our lint tests?

I think so. I only included the lints from the backend. There's quite a number of other lints to clean up that PyCharm warned about.

**Comment by willcl-ark** [comment]
> I think so. I only included the lints from the backend. There's quite a number of other lints to clean up that PyCharm warned about.

I'd be OK with this, but how does it work? I don't have that IDE and don't think you can get it's internal linter seperately?

I've also been meaning to add `ruff format` to the lints, so we always have formatted code...

**Comment by willcl-ark** [comment]
LGTM too, thanks!

**Comment by mplsgrant** [comment]
> I've also been meaning to add `ruff format` to the lints, so we always have formatted code...

@willcl-ark That seems like a great idea. I had imagined that `ruff` would be as picky as clippy right out of the box. But, I guess not!

> I'd be OK with this, but how does it work?

PyCharm has a "Try Locally" button I can press that allegedly allows me to run something called [Qodana](https://www.jetbrains.com/qodana/). When I press the button it examines everything in my local warnet project and lints it.

**Pull request merged**